Production planning backend

ScarpMetal opened the pull request:
Backend for all things related to the production planning dashboard.

ScarpMetal committed:
Backend for getting raw material and work in progress data

ScarpMetal committed:
ignore environment file

ScarpMetal committed:
Added warning to raw materials if they are running low

Also refactored some names

ScarpMetal committed:
Calculates how much of a selected (process, product) each of its ancestors can make

ScarpMetal committed:
Add total inventory to InventoryRemainingRawMaterialSerializer + cleanup for PR

ScarpMetal committed:
Removed comment

ScarpMetal committed:
Commented and refactored

ScarpMetal committed:
wording

ScarpMetal committed:
Merge branch 'staging' into production-planning-backend

mayanb commented on ics/v11/queries/production_planning.py:
```diff
@@ -0,0 +1,90 @@
+
+def get_conversion_rates(conversion_map, source):
+	# calculate the shortest paths from source to each node
+	level = 0
```
why do we need this variable if we're not using it for anything?

ScarpMetal replied:
removed

mayanb commented on ics/v11/queries/production_planning.py:
```diff
@@ -0,0 +1,90 @@
+
+	# calculate the conversion rates of each path
+	rates = {}
+	for pathKey in paths:
```
i don't think we need to calculate the conversion rate for every single ingredient type - we should do it on the fly only for the ones we actually care about

mayanb replied:
also can we calculate the rates while we're generating the path?

mayanb commented on ics/v11/queries/production_planning.py:
```diff
@@ -0,0 +1,90 @@
+					conversion_map[childKey] = {}
+	return conversion_map
+
+def get_conversion_rates(conversion_map, source):
```
if you referenced code from another library that you're not importing, make sure to cite that in a comment :)

mayanb commented on ics/v11/serializers.py:
```diff
@@ -382,6 +384,101 @@ class Meta:
+
+	def get_date_exhausted(self, item_summary):
+		# compute the rate of consumption
+		timeElapsed = timedelta(days=30).total_seconds()
```
you should make this time period a constant since it needs to be the same one you used when calculating the amount used in the queryset - right now someone could inadvertently change one of these numbers in one place and it would make things incorrect

mayanb commented on ics/v11/views.py:
```diff
@@ -591,7 +593,124 @@ def get_queryset(self):
+
+      # extract all unique ancestors of specified category (i.e raw material, work in progress, finished goods)
+      ancestorProcesses = []
+      ancestorProducts = []
```
isn't this going to mean if we use the ingredients R-MM16 and L-SUG we're going to end up looking for all items including R-MM16, R-SUG, L-MM16 and L-SUG which isn't what we want?

ScarpMetal replied:
Removed this part in the current implementation

mayanb commented on ics/v11/views.py:
```diff
@@ -591,7 +593,124 @@ def get_queryset(self):
+      ).order_by('-created_at')[:10]
+
+      # extract all unique ancestors of specified category (i.e raw material, work in progress, finished goods)
+      ancestorProcesses = []
```
if we're reusing code we should pull it out into a separate function

mayanb reviewed (changes requested):
see comments

mayanb commented on ics/v11/queries/production_planning.py:
```diff
@@ -0,0 +1,90 @@
+			)
+			for ti in task_ingredients:
+				batch_size = ti.task.items.aggregate(amount=Coalesce(Sum('amount'), 0))['amount']
+				conversion_rate = batch_size / ti.actual_amount
```
need to check if actual amount is 0 before diving by it

mayanb commented on ics/v11/serializers.py:
```diff
@@ -382,6 +384,101 @@ class Meta:
+		product_type = item_summary['creating_task__product_type']
+		return get_adjusted_item_amount(process_type, product_type)
+
+	def get_can_make(self, item_summary):
```
if the adjusted_item_amount is negative, you should cap can_make at 0 rather than telling them they can make a negative amount

mayanb commented on ics/v11/serializers.py:
```diff
@@ -382,6 +384,101 @@ class Meta:
+			conversion_rate = conversion_rates[(process_type, product_type)]
+		else:
+			conversion_rate = 0
+		amount = get_adjusted_item_amount(process_type, product_type)
```
should return get_adjusted_item_amount for each item as part of the queryset returned by the view

mayanb commented on ics/v11/views.py:
```diff
@@ -591,7 +593,124 @@ def get_queryset(self):
+          amount_used = Sum('amount'),
+        )
+
+    return queryset
```
add this code after
```
    returnobj = []
    for x in queryset:
      proc = x['creating_task__process_type']
      prod = x['creating_task__product_type']
      amt = x['amount_used']
      adj_amt = get_adjusted_item_amount(proc, prod)
      returnobj.append({'creating_task__process_type': proc, 'creating_task__product_type': prod, 'amount_used': amt, 'adj_amt': adj_amt})
    return returnobj

mayanb commented on ics/v11/views.py:
```diff
@@ -591,7 +593,124 @@ def get_queryset(self):
+      ancestorProducts = []
+      for sampleTask in sampleTasks:
+        # gets all ancestors of a task
+        ancestors = sampleTask.ancestors().filter(process_type__category=ancestor_category)
```
we shouldn't have to do this here if we're already doing this to make the conversion map - when we're going through the conversion map we can get the key's that have the process type in the category we want

ScarpMetal committed:
Moved get_adjusted_item_amount calculations from the serializer to the view

ScarpMetal committed:
Handle zero problems

ScarpMetal committed:
PR changes

ScarpMetal committed:
Reworked production planning to use one view

Now uses the endpoint ics/v11/production-planning/

ScarpMetal committed:
Removed unused variables

ScarpMetal committed:
Reduced Production Planning to a single serializer

ScarpMetal committed:
Renamed variables

ScarpMetal committed:
Not enough recent usage data feature

ScarpMetal committed:
get_queryset_info now returns an array instead of an object

The ProductionPlanning view can be shortened significantly because now it only needs to return the get_queryset_info() function

ScarpMetal committed:
Merge branch 'staging' into production-planning-backend

mayanb commented on ics/v11/queries/production_planning.py:
```diff
@@ -0,0 +1,130 @@
+# Author: Aric Hagberg (hagberg@lanl.gov)
+# Link: https://networkx.github.io/documentation/networkx-1.9/reference/generated/networkx.algorithms.shortest_paths.unweighted.single_source_shortest_path.html?highlight=single_source_shortest_path#networkx.algorithms.shortest_paths.unweighted.single_source_shortest_path
+def get_queryset_info(conversion_map, source):
+	if not source in conversion_map:
```
can we break this up into shorter helper functions? maybe 1 to get all the shortest paths and another to build up the queryset info list and one to calculate the conversion rate and the amount used

mayanb commented on ics/v11/queries/production_planning.py:
```diff
@@ -0,0 +1,130 @@
+	if not source in conversion_map:
+		return []
+
+	selected_process_type = source[0]
```
what are these 2 variables being used for?

ScarpMetal replied:
Oh I think I might have been using them in an earlier implementation, but I can remove them now

ScarpMetal committed:
Made helper functions for get_queryset_info

- changed `import datetime` to `from datetime import datetime, timedelta` :clock4:
- broke up the content of `get_queryset_info` into multiple helper functions

mayanb commented on ics/v11/serializers.py:
```diff
@@ -361,6 +363,76 @@ class Meta:
+		return futureDate
+
+	def get_warning(self, item_summary):
+		dateExhausted = self.get_date_exhausted(item_summary)
```
![image](https://user-images.githubusercontent.com/7052505/44926390-1a73b600-ad06-11e8-8aa1-2596ef8a415b.png)
right now this means if there's no data for a rm from the last 30 days even if there's 0 of it left in stock we're not going to get a warning. we should add more conditions to handle that

ScarpMetal replied:
I'm confused though, because if their amount is zero and their rate of consumption is zero, then it doesn't make sense to warn them that they are going to run out. For example, if they only made B-MT once, but don't have plans to make it again, they would be perpetually warned that they are going to run out of B-MT even though they are not actively using it.

mayanb replied:
Well I think the idea is if they’re looking at the production planning info for something that uses b-mt it’s because they want to know how much of it they can make. And if they have none of one of its ingredients on stock they should get warned about that

ScarpMetal committed:
Removed code created in this PR, but ended up becoming obsolete

ScarpMetal committed:
Handle case where there is no usage data in last 30 days, but amount is zero

mayanb commented on ics/v11/serializers.py:
```diff
@@ -616,9 +691,11 @@ def get_actual(self, goal):
 	class Meta:
 		model = Goal
-		fields = ('id', 'all_product_types', 'process_type', 'goal', 'actual', 'process_name', 'process_unit', 'process_icon', 'product_code', 'timerange', 'rank', 'is_trashed', 'trashed_time', 'userprofile_name', 'userprofile')
+		fields = ('id', 'all_product_types', 'process_type', 'goal', 'actual', 'process_name', 'process_unit', 'process_icon', 'product_code', 'timerange', 'rank', 'is_trashed', 'trashed_time', 'userprofile_name', 'userprofile', 'total_inventory_amount')
```
you probably want to remove total_inventory_amount from the fields list?

ScarpMetal committed:
Removed outdated field total_inventory_amount

Was used in a previous implementation, but is no longer used

ScarpMetal committed:
Merge branch 'staging' into production-planning-backend

ScarpMetal committed:
Updated to handle new ancestors query
